password-cli/app/Main.hs

```diff
@@ -0,0 +1,141 @@
+            <> command "scrypt" (info commandScrypt (progDesc "Scrypt operations"))
+            <> command "help" (info commandHelp (progDesc "Show command help"))
+        )
+
+commandHelp :: Parser (IO ())
+commandHelp =
+  go
+    <$> optional (argument str (metavar "COMMAND"))
+    <**> helper
+  where
+    go mCmd =
+      let args = maybe id (:) mCmd ["-h"]
+       in void $ handleParseResult $ execParserPure defaultPrefs cliOpts args
+
+commandArgon2 :: Parser (IO ())
+commandArgon2 =
+  subparser
+    ( command
+        "hash"
+        ( info
+            (pure commandHash)
+            (progDesc "hash password (via STDIN)")
+        )
+        <> command
+          "check"
+          ( info
+              (commandCheck <$> option str (metavar "HASH" <> long "hash") <**> helper)
+              (progDesc "check hashed password (via STDIN)")
+          )
+    )
+    <**> helper
+  where
+    commandHash = runHash Argon2.hashPassword
+    commandCheck hash = runCheck $ \password -> Argon2.checkPassword password $ PasswordHash hash
+
+commandBcrypt :: Parser (IO ())
+commandBcrypt =
+  subparser
+    ( command
+        "hash"
+        ( info
+            (pure commandHash)
+            (progDesc "hash password (via STDIN)")
+        )
+        <> command
+          "check"
+          ( info
+              (commandCheck <$> option str (metavar "HASH" <> long "hash") <**> helper)
+              (progDesc "check hashed password (via STDIN)")
+          )
+    )
+    <**> helper
+  where
+    commandHash = runHash Bcrypt.hashPassword
+    commandCheck hash = runCheck $ \password -> Bcrypt.checkPassword password $ PasswordHash hash
+
+commandPBKDF2 :: Parser (IO ())
+commandPBKDF2 =
+  subparser
+    ( command
+        "hash"
+        ( info
+            (pure commandHash)
+            (progDesc "hash password (via STDIN)")
+        )
+        <> command
+          "check"
+          ( info
+              (commandCheck <$> option str (metavar "HASH" <> long "hash") <**> helper)
+              (progDesc "check hashed password (via STDIN)")
+          )
+    )
+    <**> helper
+  where
+    commandHash = runHash PBKDF2.hashPassword
+    commandCheck hash = runCheck $ \password -> PBKDF2.checkPassword password $ PasswordHash hash
+
+commandScrypt :: Parser (IO ())
+commandScrypt =
+  subparser
+    ( command
+        "hash"
+        ( info
+            (pure commandHash)
+            (progDesc "hash password (via STDIN)")
+        )
+        <> command
+          "check"
+          ( info
+              (commandCheck <$> option str (metavar "HASH" <> long "hash") <**> helper)
+              (progDesc "check hashed password (via STDIN)")
+          )
+    )
+    <**> helper
+  where
+    commandHash = runHash Scrypt.hashPassword
+    commandCheck hash = runCheck $ \password -> Scrypt.checkPassword password $ PasswordHash hash
+
+runHash :: (Password -> IO (PasswordHash a)) -> IO ()
+runHash f = do
+  password <- mkPassword <$> T.hGetLine stdin
```

Comment:
It feels like it might be better to have a have the CLI work like this:

```console
$ password-cli argon2 hash
Enter password: 
<some hash>
```

After `Enter password: `, the tool should set the output to non-printing (similar to how the `passwd` tool works).

Reply:
Would this also allow usage like `cat mypass.txt | password-cli hash argon2`?
Maybe also an option like `--quiet` to not print to `stdout`?